src/code42cli/cmds/devices.py

```diff
@@ -0,0 +1,383 @@
+    required=False,
+    type=bool,
+    default=False,
+    is_flag=True,
+    help="Include device settings in output.",
+)
+@format_option
+@sdk_options()
+def list(
+    state,
+    active,
+    days_since_last_connected,
+    drop_most_recent,
+    org_uid,
+    include_backup_usage,
+    include_usernames,
+    include_settings,
+    format,
+):
+    """Outputs a list of all devices."""
+    devices_dataframe = _get_device_dataframe(
+        state.sdk, active, org_uid, include_backup_usage
+    )
+    if drop_most_recent:
+        devices_dataframe = _drop_n_devices_per_user(
+            devices_dataframe, drop_most_recent
+        )
+    if days_since_last_connected:
+        devices_dataframe = _drop_devices_which_have_not_connected_in_some_number_of_days(
+            devices_dataframe, days_since_last_connected
+        )
+    if include_settings:
+        devices_dataframe = _add_settings_to_dataframe(state.sdk, devices_dataframe)
+    if include_usernames:
+        devices_dataframe = _add_usernames_to_device_dataframe(
+            state.sdk, devices_dataframe
+        )
+    formatter = DataFrameOutputFormatter(format)
+    formatter.echo_formatted_dataframe(devices_dataframe)
+
+
+def _get_device_dataframe(sdk, active=None, org_uid=None, include_backup_usage=False):
+    devices_generator = sdk.devices.get_all(
+        active=active, include_backup_usage=include_backup_usage, org_uid=org_uid
+    )
+    devices_list = []
+    for page in devices_generator:
+        devices_list.extend(page["computers"])
+    return DataFrame.from_records(
+        devices_list,
+        columns=[
+            "computerId",
+            "guid",
+            "name",
+            "osHostname",
+            "status",
+            "lastConnected",
+            "backupUsage",
+            "productVersion",
+            "osName",
+            "osVersion",
+            "userUid",
+        ],
+    )
+
+
+def _add_settings_to_dataframe(sdk, device_dataframe):
+    macos_guids = device_dataframe.loc[
+        device_dataframe["osName"] == "mac", "guid"
+    ].values
+    rows = [
+        {"guid": guid, "macos_guid": guid in macos_guids}
+        for guid in device_dataframe["guid"].values
+    ]
+
+    def handle_row(guid, macos_guid):
+        try:
+            current_device_settings = sdk.devices.get_settings(guid)
+        except Exception as e:
+            return {
+                "guid": guid,
+                "ERROR": "Unable to retrieve device settings for {}: {}".format(
+                    guid, e
+                ),
+            }
+        current_result_dict = {
+            "guid": current_device_settings.guid,
+            "included_files": [
+                backup_set.included_files
+                for backup_set in current_device_settings.backup_sets
+            ],
+            "excluded_files": [
+                backup_set.excluded_files
+                for backup_set in current_device_settings.backup_sets
+            ],
```

Comment:
Since `.included_files` and `.excluded_files` return lists, these comprehensions end up looking like `[[:user]]` for a single backup set or `[[:user],[:user]]` for multiple backup sets. We could flatten to a single list and remove duplicates by using a set:

```
included_files = set()
for bs in current_device_settings.backup_sets:
    for path in bs.included_files:
        included_files.add(path)

current_result_dict = {
             "guid": current_device_settings.guid,
              "included_files": list(included_files)
...
```

Reply:
This is what I had here originally, but I changed it to a list of lists to better represent the case where there are multiple backup sets. For example, consider the case (of one of our customers) where there is one backup set that includes the entire user folder but excludes certain specific folders, and then a second set that includes specifically only those folders excluded by the first set. So you end up with a structure like this:

```
set 1:
    included:
         :user
    excluded:
         :user/examplefolder

set 2:
     included:
         :user/examplefolder
     excluded:
         :user/examplefolder/examplefolder2
```

If we use a set, we end up with:

```
included: :user, :user/examplefolder
excluded: :user/examplefolder, :user/examplefolder/examplefolder2
```

whereas with a list of lists, the structure is more clear:

```
included: [:user], [:user/examplefolder]
excluded: [:user/examplefolder], [:user/examplefolder/examplefolder2]
```

So in order to handle this case, I think that a list of lists is the clearest option (it is also what the current standalone script does).

Another option would be to have one row per backup set instead of one row per device, but that can also be confusing and make it more difficult to get accurate counts/filtering on the result spreadsheet (which I believe is why the standalone script also opts for a single line per device, and a list-of-lists representation for multiple backup sets).

Reply:
I'm wondering if we should split these up into two sub-commands, something like `list` and `list-backups`, where `list` just shows a single-row per device with high level status, and `list-backups` breaks it down to a row per backup set, and has _all_ the detail of each set. 

That way we don't have to worry about cramming just the right amount of information into a single row in a way that makes sense, and can leave it up to the end-user to use which output format they need depending on the use-case.

Reply:
Currently there's so little that we're doing in --include-settings that is not related to backup sets (only the full disk access status), I feel like it's fine to just have --include-settings have one line per backup set, rather than splitting everything out. I've updated it to do one line per backup set.

Reply:
I think as it is there's already a ton of information in this output, and if we're duplicating guid rows for each backup set, and also including the "regular" settings in each row, it makes it quite messy (if one device has a huge include/exclude list it can make the table _really_ wide). 

I'd prefer to separate them and have two tighter, easier-to-read outputs, which enables us to put more data in the `list-backup-config` output if we want (we're missing backup set names right now, which I'd also want to put in).

Reply:
That makes sense. I've refactored it to have a `devices list` and `devices list-backup-sets` as separate commands. Is this more what you had in mind?

Reply:
Yeah, I think that's a lot better. Since `list --include-settings` now only adds the full-disk-access state, should we separate that out into a separate function since we don't need to call `.get_settings()` on every guid in `list`?

Reply:
I will definitely remove that extraneous get_settings call. It is possible that more settings will be added at a later date, such as archive security key type or USMT migration status, at that point we can re-add the code to get the specific settings we need.

Also I seem to have introduced a bug in bulk processing that is causing multiple bulk processes to try and save their results to the same results array. I'm trying to track that down now, but it only occurs when I run the full test suite so that's a bit odd.

Reply:
Ok, `devices list --include-settings` now does not include the extraneous `.get_settings()` call, and only adds full disk access status to Mac OS devices. As more settings are added in the future, we may need to re-add that call, but for now this is all that we're reporting on.